core/src/main/java/software/amazon/awssdk/core/pagination/async/PaginatedItemsPublisher.java

```diff
@@ -0,0 +1,146 @@
+
+        ItemsSubscription(Subscriber subscriber) {
+            this.subscriber = subscriber;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                throw new IllegalArgumentException("Non-positive request signals are illegal");
+            }
+
+            outstandingRequests.addAndGet(n);
+
+            synchronized (this) {
+                if (!isTaskRunning.get()) {
+                    isTaskRunning.set(true);
+                    handleRequestsRecursively();
+                }
+            }
+        }
+
+        private void handleRequestsRecursively() {
+            if (currentPage != null && !nextPageFetcher.hasNextPage(currentPage) &&
+                    singlePageItemsIterator != null && !singlePageItemsIterator.hasNext()) {
+                subscriber.onComplete();
+                isTaskRunning.set(false);
+                return;
+            }
+
+            if (outstandingRequests.get() <= 0) {
+                isTaskRunning.set(false);
+                return;
+            }
+
+            /**
+             * Current page is null only the first time the method is called.
+             * Once initialized, current page will never be null
+             */
+            if (currentPage == null && singlePageItemsIterator == null) {
+                fetchNextPage();
+
+            } else if (singlePageItemsIterator != null && singlePageItemsIterator.hasNext()) {
+                sendNextElement();
+
+            } else if (singlePageItemsIterator != null && !singlePageItemsIterator.hasNext() &&
+                       currentPage != null && nextPageFetcher.hasNextPage(currentPage)) {
+                fetchNextPage();
+
+            // All valid cases are covered above. Throw an exception if any combination is missed
+            } else {
+                throw new IllegalStateException("Execution should have not reached here");
+            }
+        }
+
+        private void fetchNextPage() {
+            CompletableFuture<ResponseT> future = nextPageFetcher.nextPage(currentPage);
+            future.whenComplete(((response, error) -> {
+                if (response != null) {
+                    currentPage = response;
+                    singlePageItemsIterator = getIteratorFunction.apply(response);
+                    sendNextElement();
+                }
+                if (error != null) {
+                    subscriber.onError(error);
+                }
+            }));
+        }
+
+        /**
+         * Calls onNext and calls the recursive method.
+         */
+        private void sendNextElement() {
+            if (singlePageItemsIterator.hasNext()) {
+                subscriber.onNext(singlePageItemsIterator.next());
+                outstandingRequests.getAndDecrement();
+            }
+
+            handleRequestsRecursively();
+        }
+
+        @Override
+        public void cancel() {
+            outstandingRequests.set(0);
```

Comment:
Would future calls to request(n) re-activate the subscription? Should we guard against that?

Reply:
Added a isTerminated flag